More flexible/general json format specification flag

theoxo opened the pull request:
Addresses #43.

theoxo commented:
Oh also this feels a little overengineered to me... Maybe the trait isn't necessary? LMK what you think.

mlb2251 commented:
* that positional arg thing is annoying, I think we should either find a way to make it non-positional or just use Strings for simplicity instead of an enum. It looks like it's possible with enums like they reference how it is possible for "options" [here](https://docs.rs/clap/latest/clap/struct.Arg.html#method.possible_values). However it seems like the `derive` stuff is more limited. I'm not sure if it's possible to intermix `derive`d thing with non-derived things - id say either look into that or just go with strings. The non-derived approach might be like manually implementing `ArgEnum` for the type (but I haven't looked into it)

mlb2251 commented:
* for the name, maybe something like ProgramsList or something? Since we might have more stitch compatible formats too idk

mlb2251 commented:
> Oh also this feels a little overengineered to me... Maybe the trait isn't necessary? LMK what you think.

Hmmmm yeah feels a bit overengineered I think it should just be a struct method, so when someone adds a new parsing thing they just add a new variant to the enum. (Traits make sense when theres more than one struct that might impl them but I can't imagine that being the case here - if you wrote your own struct to add a new parsing method then you wont be able to loop it into `Args` - if we tried to make Args generic over some variable with this trait thatd just lead to more problems because we need to concretize that generic into a specific type before parsing and how do you do that.... basically just making people add a new variant makes sense and not doing the whole trait thing)

theoxo commented:
As of [ec3ac07](https://github.com/mlb2251/stitch/pull/68/commits/ec3ac0792ba53a500884398d1d5417035ee599ff):

Over-engineering: Addressed.
Positional arg to flag: Fixed.
Better name: Addressed.

mlb2251 commented on src/bin/compress.rs:
```diff
@@ -36,8 +34,8 @@ pub struct Args {
     pub truncate: Option<usize>,
 
     /// use dreamcoder format
```
could  you rewrite the doc comment (and list the options in it / maybe point to an example file of each type for total clarity?)

theoxo replied:
Fixed in [docs: updated docstring for fmt flag](https://github.com/mlb2251/stitch/pull/68/commits/2b7dfd51c6036b8416014e5f9f89b1e96760848c).

mlb2251 commented:
One question: I didnt see where it is you define that "dc" and "programs-list" are the two options that correspond to those variants? Is that autoinferred by the names of the enum variants?

mlb2251 commented:
Finally I'd get the OK from @CatherineWong  and @Gabe before pushing this since itll break their integrations until they change `--dc-fmt` to `--fmt=dc`

mlb2251 commented:
Other than the above 3 things it's all good to merge

theoxo commented:
> One question: I didnt see where it is you define that "dc" and "programs-list" are the two options that correspond to those variants? Is that autoinferred by the names of the enum variants?

It's autinferred from the enum names; capital letters add dashes. So "Dreamcoder" enum becomes "dreamcoder", "ProgramsList" becomes programs-list. Making it accept `--fmt=dc` would require manually overriding the `from_str()` implementation, I think

theoxo commented:
I mean if you really wanted `--fmt=dc` you could also just rename the `Dreamcoder` enum value to `Dc`, but I think that would make the code a lot less legible

mlb2251 commented:
Agreed, --fmt=dreamcoder makes sense! Just let Gabe know before you push and it's all good

CatherineWong commented:
We should probably change our `rewrite.rs` to also reflect this? If it's simple to just change the flag in our script as well? WE also copied your `--dc_fmt` naming.

CatherineWong commented:
See: https://github.com/mlb2251/stitch/blob/main/src/bin/rewrite.rs#L79 -- basically everywhere we use args.dc_fmt. I can change it too, but if its simple (we just use the exact same check.)

theoxo commented:
@CatherineWong if I understood you correctly you want the same syntax/API for `rewrite.rs` as for `compress.rs`, but you don't want to modify the existing behaviour. I implemented this in [571068c](https://github.com/mlb2251/stitch/pull/68/commits/571068c0fec2b3c29fd1bae0aa44ba621a399b8b) but don't know how to test it - could you provide an example (set of) command(s) I can use to make sure it works as intended?

CatherineWong commented:
I just glanced through, and this looks right (replacing our condition statement with the new one.) 

Our old command was `cargo run --bin=rewrite --release -- --program-file data/logo/logo_dc.json --inventions-file out/out.json --dc-fmt` which should just run (if it's not in the right format it would otherwise break.) We'd replace this I assume with the new DC Format.

gabegrand commented:
Corresponding LAPS PR is ready to go: https://github.com/CatherineWong/laps/pull/9. Confirmed that it works on this branch. @theoxo feel free to merge when you're ready.

theoxo commented:
All looks good with the example command provided by Cathy.

Merging!
